### ISSUE TITLE
Rule proposal: no-unused-assign

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#8383]
**Please describe what the rule should do:**

This rule should disallow assignments to a variable that is reassigned on every code path before being read.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
let foo = 1;

foo = 2;

doSomethingWith(foo);
```

```js
let foo = 1;

if (condition) {
  foo = 2;
} else {
  foo = 3;
}

doSomethingWith(foo);
```

```js
let foo = 1;

doSomethingWith(foo);

foo = 2;
```

To avoid conflicting with `no-unused-vars`, the rule does not report a variable if it's not used anywhere. For example, this is not reported:

```js
let foo = 1;
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a very common problem when refactoring. Fixing errors like this helps to avoid unnecessary variable mutation and makes code easier to reason about.

**Comment by platinumazure** [comment]
I think I would prefer this as a separate rule personally. The rule talks about *variables* ("vars") being unused, not variable values. In both examples, the variable is used, but a value isn't used on any code path.

**Comment by mysticatea** [comment]
I think it's one of commonly practice "a variable should be initialized at the declaration", so I don't think it's a mistake.
If we add the check, I think it should be an independent rule such as `no-unused-assign`.

**Comment by not-an-aardvark** [comment]
Thanks for the feedback. I updated this proposal to suggest a new rule rather than overloading `no-unused-vars`.

In response to @mysticatea, I agree that it's common practice for variables to be initialized at the declaration, but I think the purpose of that guideline is to make the value of the variable clear. Setting an unused value and reassigning the variable later on would make the code more confusing.

**Comment by not-an-aardvark** [comment]
Closing because there wasn't consensus on this.
